timely/src/dataflow/channels/pushers/exchange.rs

```diff
@@ -18,42 +19,60 @@ impl<T: Clone, D, P: Push<Bundle<T, D>>, H: FnMut(&T, &D)->u64>  Exchange<T, D,
     pub fn new(pushers: Vec<P>, key: H) -> Exchange<T, D, P, H> {
         let mut buffers = vec![];
         for _ in 0..pushers.len() {
-            buffers.push(Vec::with_capacity(Message::<T, D>::default_length()));
+            buffers.push(Vec::with_capacity(Message::<T, D>::minimal_length()));
         }
         Exchange {
             pushers,
             hash_func: key,
+            max_len: vec![0; buffers.len()],
             buffers,
             current: None,
         }
     }
     #[inline]
-    fn flush(&mut self, index: usize) {
+    fn flush(&mut self, index: usize, maybe_shrink: bool) {
         if !self.buffers[index].is_empty() {
+            let old_len = self.buffers[index].len();
             if let Some(ref time) = self.current {
                 Message::push_at(&mut self.buffers[index], time.clone(), &mut self.pushers[index]);
             }
+            if maybe_shrink {
```

Comment:
@antiguru , I'm not sure I understand the shrinking logic here.  Let's say, the buffer has grown to a large size `N`, so `capacity=max_len=N`.  In this case`flush(true)` will set `max_len` to 0, but will not deallocate the buffer because `capacity > 2 * max_len` doesn't hold.  The buffer will only get deallocated next time`flush(true)` is called assuming `capacity` is small.  Is this the desired behavior?